### ISSUE TITLE
Changing tags for armor recipes

### ISSUE THREAD

**Issue by earthsend** [Arx-Game/arxcode#113]
**Is your feature request related to a problem? Please describe.**
Right now, you have to set certain things manually beyond basevalue like penalties to dodge and fatigue when you create recipes for armor.  This ends up being a bit of a chore, as the code snippets containing the various tags you can set like 'baseval' are spread out in the codebase instead of kept in a single location.

**Describe the solution you'd like**
A  way to make this simple for the database as a whole is just to declare variables that can be used as tags. For instance,  a 'heavy_armor' tag  could add a +10 difficulty to dodge and +5 difficulty to fatigue rolls, and so on. 

**Describe alternatives you've considered**
A list of possible tags for clothing, leatherworking, weaponsmithing, etc. recipes to be placed in a text  or reference file might also solve this problem.

**Comment by TehomCD** [comment]
Tags would be one approach, but I think how I'd do this would be at the typeclass level - if you're already going to have hard-coded determinations of what the tags should do, then it'd probably make sense to have those be properties that'd be set per typeclass. I don't really like how I had most of them set values as Attributes in the crafted item - it's unnecessary since they can all be derived by recipe, and is a relic from ages ago before recipes existed. The drawbacks to this approach (typeclasses rather than tags) is it'd lack the flexibility of tags if you wanted to have armor that could transform type (changing typeclasses can be pretty hairy), but I don't really see that as a common use, and somewhat more clutter in the hardcoded charfield of recipe type to typeclass. If I did it over again I'd probably make the charfield the full typeclass path and then have some post-creation method for per-typeclass initial crafting setup, but eh.

I don't like the current slot system, which is why baseval is used at all - ideally, slots would exist in the database with certain values and then what's currently baseval would be derived from the combination of quality/recipe level and a value for that particular slot. Same with all the various rules for weapons, which I'll almost certainly change eventually.

I don't think I'd be a fan of placing data that might need to be modified in a flatfile or hardcoded in a module - in my opinion it's a lot easier to modify database tables via django admin or with a `@py` call than needing to either keep flat values in version control or have to SSH in to change a configuration file.

**Comment by earthsend** [comment]
Yeah, something that's easier to configure via django admin would probably work out best for this. 

Most of what I find hard to configure with the recipes is just knowing _what_ I need to put on the field. For instance, it took me a good hour to find out that medium wpn is the default weapon skill for a weapon recipe.

**Comment by Zironic** [comment]
The way I would do this is that I would implement a template class that would be almost identical to the recipe class which would have a many to many relationship with the recipes. Then the stats of any given recipe would be the sum of the stats of all its templates and its own stat. Then if you wanted to change all derived items you'd simply change the template in Django Admin.

**Comment by TehomCD** [comment]
Like `RecipeAttribute` or something? That's a possibility too, probably with what the attribute does being some integer choices field or whatever.

**Comment by Zironic** [comment]
Well, there are different ways to do it depending on how naturalized you want it to be. For a different game I designed an item system where each item was made up from components where each component could have a material kind of like Tinkerer's Constructs from Minecraft.

However in this case, if what you're after is being able to define recipe behaviors, then I'd probably just go with recipe templates (or any other equivalent term like archetypes etc). Conveniently you can use the same tool to at the same time standardize how slots work as well as standardize material stats since you could then have the recipe derive its penalties and base stats from its material and slot instead of having it set individually. The downside is that the DB queries get somewhat complicated since to display the stats of any given item, the game needs to grab the recipe and then grab the material and the recipe types in order to derive the final stats.

I hope that all made sense.

**Comment by earthsend** [comment]
For the record, I forgot which issue it was that referenced the use of armor being synthetic HP, but that could be a very good solution to the approach of cloth armor vs leather armor vs metal armor. So if how armor is handled changes, this could be something that can be done along with it to make recipes easier/better to set across games.
